### ISSUE TITLE
Moving to another DNS provider

### ISSUE THREAD

**Issue by wdhdev** [is-a-dev/register#14153]
###### (reference #13995)

---

# Alternatives

## [Cloudflare](https://www.cloudflare.com/application-services/products/dns/)
- **Pros**
	- Free, up until certain limits
	- Cloudflare proxy option
	- Caching via Cloudflare's global CDN (when proxy is enabled)
	- Faster DNS deployment
	- Built-in DDoS protection for subdomains (when proxy is enabled)
	- NS record support for subdomains
	- Free email forwarding (can setup support@is-a.dev, security@is-a.dev, etc)
	- Cloudflare Zero Trust
		- Tunnels
- **Cons**
	- Record limit, although can be increased by contacting support I believe
	- Not great support

## [Bunny.net](https://bunny.net/pricing/dns/)
- **Pros**
	- Good support
	- Custom/whitelabel nameserver support (e.g. ns1.is-a.dev & ns2.is-a.dev show as the nameserver records, but Bunny.net is the backend)
	- DNS query logs and statistics
	- Importing zone file
- **Cons**
	- Paid
	- Limit of 20 million queries a month, before it's paid
![image](https://github.com/is-a-dev/register/assets/87287585/39748467-807e-44c4-92a0-55b91e563a72)
    - No NS record support (also has limited records types available, compared to Cloudflare)

---

## Overall cons
- Moving CI over could be a hassle, however using [DNSControl](https://dnscontrol.org) could make it easier. (used with Open Domains and Free Domains)
- URL records, not sure how we would manage to implement these, but we could figure out a way.

**Comment by wdhdev** [comment]
> I can write an URL redirector in Python (however would not be a preferable solution, the more preferable one would be you writing one on JS or see the next line)
however I think you can also deprecare URL records (maybe send an email to people with URL records to tell them to remove before X?)

We can do this, and just setup a VPS for it to run on.

> the number of records we have now has already surpassed the limit of [1000 for free and 3500 for paid](https://community.cloudflare.com/t/how-to-add-more-dns-record-quota-to-a-pro-account/586515).

Cloudflare support can help with this. JS.ORG exceeds that limit I'm pretty sure, and they are on the free plan iirc.

**Comment by wdhdev** [comment]
@zvdxcsite If we did self host our DNS, it would most likely be with a more advanced and more well-known project.

**Comment by NullyIsHere** [comment]
Also it should be a really good one and good server cause it would probably crash.

**Comment by wdhdev** [comment]
@phenax After some research (and testing), I think if we did self host our DNS it would be quite beneficial, the only thing is we would need to get at least 2 VPS' with 100% uptime (in 2 different locations, maybe one in the US and another in EU?). Another thing we would need is API support.

After testing [Technitium DNS Server](https://technitium.com/dns/), I've found it would probably would work very well. I have personally tried running this and the DNS propagation is extremely fast, and the software itself supports many different record types (including NS) and is super easy to use and understand. With little to no experience in hosting DNS servers, I managed to quite easily host Technitium and setup a DNS zone and got it all up in running with-in <30 minutes. It also has a HTTP API, which would be perfect for our project.

The only main downsides to this is basically learning the API, although I wouldn't think it would be too hard and possibly DNS resolution times would be a bit slower than using a provider like Namecheap or Cloudflare. The software requirements are extremely low, so cost would be very low to run (possibly cheaper than CPanel?). As we use Namecheap as our domain registrar, registering nameservers is very easy.

The reason for 2 VPS' would be one root nameserver and then a secondary DNS server so in the event the root nameserver goes down, the secondary nameserver can still resolve the DNS zone.

![ScreenShot1](https://github.com/is-a-dev/register/assets/87287585/1f05f1f6-3bef-4877-a35a-3c54b19672bd)

**Comment by NullyIsHere** [comment]
Seems cool, @wdhdev since I read that i started to work on something for URL records, its almost finish.

**Comment by wdhdev** [comment]
@NullyIsHere Sounds good, let me know when you're finished!

**Comment by NullyIsHere** [comment]
Status update: Working on a script to detect URL records, where and in what domain.

**Comment by phenax** [comment]
A few points to consider here:
- URL redirection is a separate problem to DNS so it can be dealt with separately. So instead of trying to migrate everything at once, we can simplify things a lot by solving 1 problem at a time.
- It needs to have preferably no (or at least very high) limits on the number of dns records (and redirections) or rate-limits on the api to manipulate those records.
- Github cannot be the source of truth for data for either of them. I.e. no fetching records directly from github. The records must be saved somewhere else in a more efficient/indexed form.
- Prefer fixed pricing for our solutions. Since it's more predictable.
- Has to support at least all of our current features.
- Scaling is a pretty messy problem and I have no idea when scaling becomes a concern for dns servers. If we're going with a self-hosted solution, we'll need to perform extensive load testing and we'll need really good monitoring tools.
- If we're relying on a managed service, customer service needs to be really good or else we're stuck dealing with problems we have no control over.
- Security will be hard to validate but we'll need to keep an eye on it and be as conservative about access as possible.


@wdhdev, [Technitium](https://technitium.com/dns/) looks great. Appreciate the thorough research! I'll experiment with it a bit this week.
If that doesn't work out bunny.net seems like a decent managed alternative (zone management, not the scriptable dns). The pricing is not fixed unfortunately but it seems reasonable enough.

> No NS record support (also has limited records types available, compared to Cloudflare)

It seems like bunny supports it [here](https://docs.bunny.net/reference/dnszonepublic_addrecord) but I'm not sure if there are any implicit restrictions on it.

**Comment by wdhdev** [comment]
@phenax I've done some testing with Technitium and so far, it seems very good. It even allows you to host your own local domains and TLDs if wanted. It has easy to setup secondary nameservers as well (only a few clicks to setup). It also has in-depth statistics.

Thanks for your response!

**Comment by wdhdev** [comment]
@MaskDuck @phenax For hosting, uptime and reliability this is what I personally think we would/should do:
- Main server (ns1.is-a.dev)
  - Hosted on a VPS provider like [OVHcloud](https://us.ovhcloud.com) in the United States
- Secondary server (ns2.is-a.dev)
	- Hosted on the same provider, but somewhere else in the EU, like Germany or the UK.
- Possible 3rd server (ns3.is-a.dev) (*optional*)
  - Also hosted on the same provider, but in Australia.
- And so on...

This way, if the main nameserver, `ns1.is-a.dev` goes down, `ns2.is-a.dev` and `ns3.is-a.dev` (if it exists), will just pickup the load immediately causing no downtime for DNS queries.

TL;DR; Have 2 or 3 servers in different regions to minimise loading times for users in different regions and to avoid downtime if a server goes down, therefore increasing reliability.

The reason why I recommended OVHcloud is because they have a high SLA for most services and notice in advance for maintenance.

**Comment by wdhdev** [comment]
> I'd recommend you two host two different nameservers on two different platforms

Same provider, different datacenters will be good enough for our needs, in my opinion. Also, 2 nameservers should be enough, the main one in the US, a secondary one in the EU.

**Comment by NullyIsHere** [comment]
Germany could be a good option as it could cover parts of the East

**Comment by wdhdev** [comment]
> Germany could be a good option as it could cover parts of the East

Yeah, I was thinking Germany.

**Comment by wdhdev** [comment]
> The server that used to run hosting and manage site is Located in London, England that could also become part of this as it only running the discord bot and open-domains plus it has a 25GB uplink

We will need DDoS protection & stuff. Unfortunately, we probably won't be able to use your server unless Phenax wants to of course.

**Comment by phenax** [comment]
I might be wrong but from what I can tell, Technitium only works on disk. Which means managing multiple instances with that will be a shit show. Both on how we update dns records across instances and how we scale instances.

We'll need a common data source that lives outside disk. This is because killing an instance should not mean all the records are gone and we now have the headache of migrating the records on disk to the new instance before that. Which we may not be able to do reliably. So we'll need 2 instances running the dns server and 1 db instance with backups.

If the caching strategy isn't good enough, a ddos attack could potentially overload the db. But the chance of that is pretty slim and can be dealt with later with replication. We'll have a better idea after some stress testing and tweaking our dns solution a bit.

[PowerDNS](https://www.powerdns.com/) seems like a pretty strong alternative that supports [external backends for storage](https://doc.powerdns.com/authoritative/backends/index.html) and has decent caching which means we can deploy multiple instances of this and scale it without much hassle. Also has an api and a dashboard. Will try with this out later.

Let me know what you all think.

**Comment by wdhdev** [comment]
@phenax You can just setup 2 or more instances, run the zone on a main instances, then just configure secondary zones on the rest after adding NS records for them in the primary zone.

The zone is automatically synced to secondary servers.

I have my own setup using Technitium including a secondary nameserver and I'm happy to show you how I have it configured. Me and @MaskDuck were just doing some testing with it. I can hop in a Discord call with you if you want, anytime today, just DM me or ping me in the support server.

I have seen PowerDNS before, however it seems a bit conplicated to setup, at least for me. Also, in my opinion, it is better to use the method Technitium has for syncing secondary nameservers as it doesn't rely on a external backend, so that way there is multiple copies of the zone which are all synced every 30 seconds or so with the primary nameserver and there is no issues if the primary nameserver goes down as the secondary nameserver has a copy of the zone.

**Comment by wdhdev** [comment]
@phenax I've sent you a message on Discord showing you how to manage Technitium with multiple instances and some recommended options to enable, with screenshots. It is super easy to setup and to scale.
The main downside is just having a login for each instance, other than that, it is super easy to scale.

**Comment by andrewstech** [comment]
> > The server that used to run hosting and manage site is Located in London, England that could also become part of this as it only running the discord bot and open-domains plus it has a 25GB uplink
> 
> We will need DDoS protection & stuff. Unfortunately, we probably won't be able to use your server unless Phenax wants to of course.

My server is now equipped with DDOS protection, Upgraded my plan last month. Also all servers must have an IPV6 address otherwise AAAA records won’t be resolved

**Comment by wdhdev** [comment]
> My server is now equipped with DDOS protection, Upgraded my plan last month.

Alright then, we can most likely configure it as a secondary nameserver, the more nameservers the better 😆 

I might consider hosting one as well.

> Also all servers must have an IPV6 address otherwise AAAA records won’t be resolved

I've just tested my Technitium instance, which does not have an IPv6 assigned and AAAA records do in fact resolve. You can see the output here:
```
$ dig ipv6.test.wdh.gg AAAA +noall +answer
ipv6.test.wdh.gg.       3600    IN      AAAA    2001:4860:4860::8888
```

![image](https://github.com/is-a-dev/register/assets/87287585/a7016e34-e363-4124-b357-1da5105c9cf5)

**Comment by andrewstech** [comment]
> One problem i thought about is that ns cant be n1.is-a.dev as to resolve that needs to resolve n1.is-a.dev, etc. So it should be n1.wdh.gg oor similar

We can use glue records to resolve this. They are configured on the domain before the nameserver. They sort of sit on top

**Comment by wdhdev** [comment]
> One problem i thought about is that ns cant be n1.is-a.dev as to resolve that needs to resolve n1.is-a.dev, etc. So it should be n1.wdh.gg oor similar

Glue records is the answer as @andrewstech said before. These are configured at the registrar level, most (if not all) major registrars like Namecheap support them.

**Comment by NullyIsHere** [comment]
Thanks for the explanarion.

**Comment by wdhdev** [comment]
@NullyIsHere Hey, are you able to join our [Discord server](https://discord.gg/is-a-dev-830872854677422150) to discuss the URL redirection script?

**Comment by NullyIsHere** [comment]
Yeah, ill join now.

**Comment by phenax** [comment]
> The zone is then automatically synced to secondary servers.

That sounds alright. I just have a painful past of debugging systems with multiple sources of truth. Shouldn't be a huge deal either way.


> I have seen PowerDNS before, however it seems a bit complicated to setup, at least for me.

This shouldn't be a factor since we won't be managing our infrastructure by hand every time.


> that way if the main Technitium instance goes down (or if using PowerDNS, the external backend), it will not disrupt traffic at all

Not an issue with an external backend because they can be scaled independently of the instances with a read-only replica.


@wdhdev The problem I see with Technicium is that it doesn't seem like a production-grade DNS solution. They seem to advertise themselves as a home/org network solution and ad blocker like pi-hole which doesn't give me any confidence. Feels like force a fit. Let me know what you think.

But yeah, I guess either on-disk or external should work fine for us. Might be worth looking into more solutions.

**Comment by wdhdev** [comment]
Thanks for your response, I do agree it is more personal use and meant for more smaller networks, so maybe it isn't the solution for us.

PowerDNS does seem like it is advertised more for bigger applications like ours. I might spin up a PowerDNS instance later today and send a few screenshots either here or on Discord.

**Comment by andrewstech** [comment]
> That sounds alright. I just have a painful past of debugging systems with multiple sources of truth. Shouldn't be a huge deal either way.
> 
> >

If we did take the PowerDNS route then we can use the CI from **Open Domains**. As that uses [DNSControl](https://dnscontrol.org/)  to automate record deployments as it has support for PowerDNS.

CI: https://github.com/open-domains/register/blob/main/.github/workflows/publish.yml

Config: https://github.com/open-domains/register/blob/main/dnsconfig.js

Would be very easy to adjust to is-a.dev file structure

**Comment by wdhdev** [comment]
> PowerDNS does seem like it is advertised more for bigger applications like ours. I might spin up a PowerDNS instance later today and send a few screenshots either here or on Discord.

I got a bit lazy and forgot about this. I tried installing a PowerDNS instance with PowerDNS-Admin however for whatever reason I could not get it working for whatever reason (@Stef-00012 and @CuteDog5695 can confirm this as they were watching my screenshare on Discord). @phenax if you could try spin up an instance that might be better.

**Comment by phenax** [comment]
@wdhdev I have a test setup for powerdns with pdns-admin. Pushed it in an [infra repo](https://github.com/is-a-dev/infrastructure) (private for now). We'll use that as a base for our infra.

**Comment by wdhdev** [comment]
@phenax Oh, amazing. I couldn't get PowerDNS-Admin working myself so that's good. Are you able to give me a login to the server possibly? I'd like to run some of my own tests on one of my test domains.

**Comment by wdhdev** [comment]
@phenax Any updates on this? Is PowerDNS looking like the way to go for is-a.dev? It looks good to me so far.

**Comment by AlexWinder** [comment]
> > My server is now equipped with DDOS protection, Upgraded my plan last month.
> 
> Alright then, we can most likely configure it as a secondary nameserver, the more nameservers the better 😆
> 
> I might consider hosting one as well.
> 
> > Also all servers must have an IPV6 address otherwise AAAA records won’t be resolved
> 
> I've just tested my Technitium instance, which does not have an IPv6 assigned and AAAA records do in fact resolve. You can see the output here:
> 
> ```
> $ dig ipv6.test.wdh.gg AAAA +noall +answer
> ipv6.test.wdh.gg.       3600    IN      AAAA    2001:4860:4860::8888
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/87287585/340666385-a7016e34-e363-4124-b357-1da5105c9cf5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY5NTQxNzYsIm5iZiI6MTcyNjk1Mzg3NiwicGF0aCI6Ii84NzI4NzU4NS8zNDA2NjYzODUtYTcwMTZlMzQtZTM2My00MTI0LWIzNTctMWRhNTEwNWM5Y2Y1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIxVDIxMjQzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViYmFlNDlhYzM5MjA4NTc4MzE3NjM3NmIzYWM5NDdlN2YwZGViNTRjODczNzhjNmNhOTdiNjA1OTVhOGZiNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.L93Zjgkoh6EhkDZvj0qAHA3OtvrNnXiV-AoW6gWL8xM)

I'm new to this repo, but impressed with the idea. I was browsing over how it works and stumbled upon this issue and felt like I could offer some knowledge with one thing which was said as I was in this exact situation on a project previously.

The DNS server would be required to have both an IPv4 and IPv6 address not only to resolve AAAA records, but instead to enable requests from IPv6 enabled connections to be correctly resolved. If your request was made over IPv4 then you will be able to resolve an AAAA record - even if your machine doesn't have IPv6 and is unable to even use the returned RR. So enabling IPv6 is more to do with allowing DNS lookups from clients which either use IPv6 as a priority, or only use IPv6 - in this scenario if your DNS server has no IPv6 enabled then no results will resolve for them.

Using `dig` you can force an IPv6 request with the `-6` flag.

For example, below has both IPv4 and IPv6 enabled, but the DNS server only responds to IPv4 requests.

A request to AAAA of www.google.com forced over IPv4 returns a result:

```console
root@shared-web:~# dig www.google.com AAAA -4

; <<>> DiG 9.18.28-0ubuntu0.22.04.1-Ubuntu <<>> www.google.com AAAA -4
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 45609
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;www.google.com.                        IN      AAAA

;; ANSWER SECTION:
www.google.com.         138     IN      AAAA    2a00:1450:4026:805::2004

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Sat Sep 21 21:39:55 UTC 2024
;; MSG SIZE  rcvd: 71
```

The same request forced over IPv6 which isn't used on my local DNS does not resolve anything because it doesn't listen/use IPv6 for its resolving:

```console
root@shared-web:~# dig www.google.com AAAA -6
;; communications error to ::1#53: connection refused
;; communications error to ::1#53: connection refused
;; communications error to ::1#53: connection refused

; <<>> DiG 9.18.28-0ubuntu0.22.04.1-Ubuntu <<>> www.google.com AAAA -6
;; global options: +cmd
;; no servers could be reached

root@shared-web:~# nano /etc/resolv.conf
root@shared-web:~# dig www.google.com AAAA -6
;; communications error to ::1#53: connection refused
;; communications error to ::1#53: connection refused
;; communications error to ::1#53: connection refused

; <<>> DiG 9.18.28-0ubuntu0.22.04.1-Ubuntu <<>> www.google.com AAAA -6
;; global options: +cmd
;; no servers could be reached
```

**Comment by phenax** [comment]
> Any updates on this? Is PowerDNS looking like the way to go for is-a.dev? It looks good to me so far.

@wdhdev Aah got tangled up with other stuff. I'll try to get some infra set up by next week.

@is-a-dev/maintainers, Could someone open up a draft PR setting up [DNSControl for powerdns](https://docs.dnscontrol.org/provider/powerdns) here for testing? It doesn't need to handle every case yet, just for pushing all our records for load testing.

**Comment by wdhdev** [comment]
@andrewstech can you do that?

**Comment by wdhdev** [comment]
I've opened #16286.

@phenax The test instance is returning 500 internal server errors, so I cannot use it.

**Comment by wdhdev** [comment]
@phenax As we are not fully set on moving to PowerDNS yet, I have opened a post on Cloudflare's Community forum, just to see if they could help us out, as Cloudflare DNS would be infinitely more reliable and easier to work with too.

It would also bring the features of Cloudflare's proxy which is good.

https://community.cloudflare.com/t/moving-subdomain-service-to-cloudflare-dns/716357

**Comment by wdhdev** [comment]
Update on Cloudflare, turns out they have an OSS sponsorship application, I submitted a request there a couple days ago. Still waiting on a response.

**Comment by wdhdev** [comment]
Cloudflare just announced a new OSS sponsorship program called [Project Alexandria](https://x.com/CloudflareDev/status/1839652683780636743), so I have just reapplied is-a.dev on that new form.

The main thing I requested on the form is a DNS record limit of around 25,000 as we exceed the 1,000 record limit on their free plan and 3,500 on their paid plans. If we do receive this sponsorship, we will likely move is-a.dev to Cloudflare.

**Comment by DEV-DIBSTER** [comment]
> Contributor

Agreed with MaskDuck, I doubt there's not that many other projects that need unlimited DNS records besides projects like us, and our type of project is already on Cloudflare as with .js.org.

**Comment by wdhdev** [comment]
I didn't want to be too greedy regarding the record cap. We will be able to get it increased when we hit that cap anyway.

**Comment by wdhdev** [comment]
@phenax We have good news. Cloudflare has accepted the sponsorship request. I am still working with them to get the domain added to Cloudflare, as they asked me to add the domain to Cloudflare, which is not possible as Cloudflare blocks adding domains listed on the PSL as it considers them a TLD. Basically, I'm still working with Cloudflare on getting the domain added and the upgrades applied.

I will add the domain to my developer Cloudflare account (not personal one) and will add you to the zone once it is added (please send me your Cloudflare email in Discord DMs). We will likely move the is-a.dev zone to a dedicated Cloudflare account in future, however I will just have it in my developer account for now for ease. I will also either need Namecheap access to the domain, or you will need to update the nameservers at some point.

@andrewstech Are you able to change #16286 to use Cloudflare, instead of PowerDNS, thank you.

**Comment by wdhdev** [comment]
More good news, I have contacted a DNS engineer at Cloudflare through one of my connections and they got the is-a.dev domain added as a zone to my Cloudflare account.

I have emailed the sponsorship person at CF about this and am awaiting them to upgrade the plan. 

@phenax Please DM me your Cloudflare email on Discord so I can invite you. When the upgrades are applied to the zone I'll send you the NS records and we can move everything over. I'll remain in touch with the engineer so we can ensure everything gets moved over without issue.

**Comment by wdhdev** [comment]
@phenax The CI PR has been completed, we are fully ready to transfer to Cloudflare, we just need the NS records updated and for the sponsorship to be applied, which can only happen once we get the NS records updated.

I've sent you the new NS records and DS record for is-a.dev, please update them ASAP.

**Comment by phenax** [comment]
That's awesome! Great job @wdhdev! We can schedule migration for tomorrow.

> we are fully ready to transfer to Cloudflare, we just need the NS records updated and for the sponsorship to be applied, which can only happen once we get the NS records updated.

I may be misunderstanding this but does this mean the records can only be migrated to the zone after the NS record is reflected and the sponsorship is approved? If so, is there a way to avoid any expected downtime here by doing this progressively?

**Comment by wdhdev** [comment]
@phenax We've been approved, we just need to update the NS records to get the actual sponsored plan. We should be able to do this with little downtime hopefully.

**Comment by wdhdev** [comment]
@phenax We've been approved, we just need to update the NS records to get the actual sponsored plan. We should be able to do this with little downtime hopefully. 

I've already alerted everyone in the Discord Server of the migration, if possible can we just do it ASAP?

**Comment by wdhdev** [comment]
Yep, can do. Also, if possible please check Discord DMs once more.

**Comment by wdhdev** [comment]
@phenax Please check your email regarding Namecheap.
